test/lib/tls_certificates/test_ca_client.py

```diff
@@ -0,0 +1,131 @@
+
+        self.harness.add_relation_unit(relation_id, 'easyrsa/0',
+                                       {'ingress-address': '192.0.2.2'})
+
+        self.assertTrue(len(receiver.observed_events) == 1)
+        self.assertIsInstance(receiver.observed_events[0], ca_client.CAAvailable)
+
+    def test_request_server_certificate(self):
+        relation_id = self.harness.add_relation('ca-client', 'easyrsa')
+        self.harness.update_relation_data(
+            relation_id, 'myserver/0', {'ingress-address': '192.0.2.1'})
+
+        self.harness.add_relation_unit(relation_id, 'easyrsa/0',
+                                       {'ingress-address': '192.0.2.2'})
+        rel = self.harness.charm.model.get_relation('ca-client')
+
+        example_hostname = 'myserver.example'
+        sans = [example_hostname, '192.0.2.1']
+        self.ca_client.request_server_certificate(example_hostname, sans)
+
+        server_data = rel.data[self.harness.charm.model.unit]
+        self.assertEqual(server_data['common_name'], example_hostname)
+        self.assertEqual(server_data['sans'], json.dumps(sans))
+
+        # Simulate a change and make sure it propagates to relation data correctly.
+        new_example_hostname = 'myserver1.example'
+        new_sans = [new_example_hostname, '192.0.2.10']
+        self.ca_client.request_server_certificate(new_example_hostname, new_sans)
+        self.assertEqual(server_data['common_name'], new_example_hostname)
+        self.assertEqual(server_data['sans'], json.dumps(new_sans))
+
+    def test__on_relation_changed(self):
+
+        class TestReceiver(framework.Object):
+
+            def __init__(self, parent, key):
+                super().__init__(parent, key)
+                self.observed_events = []
+
+            def on_tls_config_ready(self, event):
+                self.observed_events.append(event)
+
+        receiver = TestReceiver(self.harness.framework, 'receiver')
+        self.harness.framework.observe(self.ca_client.on.tls_config_ready, receiver)
+
+        relation_id = self.harness.add_relation('ca-client', 'easyrsa')
+        self.harness.update_relation_data(
+            relation_id, 'myserver/0', {'ingress-address': '10.209.240.176'})
+
+        self.harness.add_relation_unit(relation_id, 'easyrsa/0',
+                                       {'': '192.0.2.2'})
+
+        # Load the sample relation data from a file. The certificates and a key
+        # were generated once for the purposes of creating an example.
+        # They are not used anywhere in a production or test system.
+        ca_client_data = yaml.safe_load(
+            (Path(__file__).parent / 'ca_client_test_data.yaml').read_text())
+
+        self.harness.update_relation_data(relation_id, 'easyrsa/0', ca_client_data)
+
+        self.assertTrue(len(receiver.observed_events) == 1)
+        self.assertIsInstance(receiver.observed_events[0], ca_client.TlsConfigReady)
+
+        self.assertIsInstance(self.ca_client.ca_certificate,
```

Comment:
It seems odd to depend on an instance of a private type here.

Reply:
An alternative I suggest then is checking some properties of certs and keys based on the public properties that those instances expose.

Reply:
https://github.com/canonical/operator/pull/210/commits/7407ab5ebd35c1cf52547f399c447e0a6f758bac